### PR TITLE
Finalize Dev Null for Release

### PR DESCRIPTION
**Pull request by johnny-schmidt**
## What
Finalization of the DevNull connector migration

* supports OSS and CLOUD specs
* come clean-up around naming
* migrated to the new bulk-cdk connector plugin
* incorporates @edgao's integration test framework

DO NOT RELEASE UNTIL
* @stephane-airbyte 's merge of oss+cloud is done
* @edgao 's test work complete
* we figure out what is going on with the integration tests in the plugin 
* probably one more quick sanity test to make sure I didn't break anything

### CODE REVIEW

File: airbyte-integrations/connectors/destination-dev-null/README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
-# Destination Dev Null
```
Comment (johnny-schmidt) [inline airbyte-integrations/connectors/destination-dev-null/README.md:1]: Ignore this, I should just pick up whatever @stephane-airbyte did for the readme

File: airbyte-integrations/connectors/destination-dev-null/metadata.yaml [hunk 2]
```diff
@@ -1,8 +1,8 @@
 data:
-  connectorSubtype: file
+  connectorSubtype: unknown
```
Comment (johnny-schmidt) [inline airbyte-integrations/connectors/destination-dev-null/metadata.yaml:2]: note: double-check how @stephane-airbyte resolved this (and also make sure our versions align)

File: docs/integrations/destinations/dev-null.md [hunk 3]
```diff
@@ -8,7 +8,8 @@ The Airbyte `dev-null` Destination. This destination is for testing and debuggin
   <summary>Expand to review</summary>
 
 | Version | Date       | Pull Request                                             | Subject           |
-| :------ | :--------- | :------------------------------------------------------- | :---------------- |
```
Comment (johnny-schmidt) [inline docs/integrations/destinations/dev-null.md:11]: TODO: fix this

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt [hunk 4]
```diff
@@ -4,7 +4,10 @@
 
 package io.airbyte.cdk.test.util
 
+<<<<<<< HEAD
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt:7]: fixed this

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/message/DestinationMessageDeserializer.kt [hunk 5]
```diff
@@ -26,8 +26,7 @@ class DefaultDestinationMessageDeserializer(private val messageFactory: Destinat
             val airbyteMessage = Jsons.readValue(serialized, AirbyteMessage::class.java)
             return messageFactory.fromAirbyteMessage(airbyteMessage, serialized)
         } catch (t: Throwable) {
-            // TODO: REMOVE THIS BEFORE RELEASE!
-            throw RuntimeException("Failed to deserialize AirbyteMessage: $serialized", t)
+            throw RuntimeException("Failed to deserialize AirbyteMessage: $serialized")
```
Comment (stephane-airbyte) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/message/DestinationMessageDeserializer.kt:29]: this should probably still be removed

File: airbyte-integrations/connectors/destination-dev-null/README.md [hunk 6]
```diff
@@ -1,6 +1,6 @@
...
+# End-to-End Testing Destination
```
Comment (stephane-airbyte) [inline airbyte-integrations/connectors/destination-dev-null/README.md:1]: bad merge? I think we want to keep the DevNull name, no?

File: airbyte-integrations/connectors/destination-dev-null/build.gradle [hunk 7]
```diff
@@ -1,18 +1,26 @@
...
 }
 
+// Uncomment to run locally
+//run {
+//    standardInput = System.in
+//}
+
 dependencies {
-    implementation project(':airbyte-integrations:connectors:destination-e2e-test')
+    testImplementation platform('org.testcontainers:testcontainers-bom:1.19.8')
```
Comment (stephane-airbyte) [inline airbyte-integrations/connectors/destination-dev-null/build.gradle:25]: this seems very heavy handed

File: airbyte-integrations/connectors/destination-dev-null/metadata.yaml (hunk 2 above)
Comment (stephane-airbyte) [reply airbyte-integrations/connectors/destination-dev-null/metadata.yaml:2]: (current version is 1.20.1. Please bump it)

File: airbyte-integrations/connectors/destination-dev-null/src/test-integration/kotlin/io/airbyte/integrations/destination/dev_null/DevNullBasicFunctionalityIntegrationTest.kt [hunk 8]
```diff
@@ -26,4 +26,9 @@ class E2eBasicFunctionalityIntegrationTest :
     override fun testBasicWrite() {
         super.testBasicWrite()
     }
+
+    @Test
```
Comment (stephane-airbyte) [inline airbyte-integrations/connectors/destination-dev-null/src/test-integration/kotlin/io/airbyte/integrations/destination/dev_null/DevNullBasicFunctionalityIntegrationTest.kt:30]: are the overrides necessary for the tests to run?

File: airbyte-integrations/connectors/destination-dev-null/src/test/java/io/airbyte/integrations/destination/dev_null/DevNullDestinationTest.java
Comment (stephane-airbyte) [inline airbyte-integrations/connectors/destination-dev-null/src/test/java/io/airbyte/integrations/destination/dev_null/DevNullDestinationTest.java:1]: we can't keep the old integration tests around and run the new destination against them?

**Review by edgao — APPROVED** [review]
lgtm after comments are all resolved

File: airbyte-integrations/connectors/destination-dev-null/src/test-integration/kotlin/io/airbyte/integrations/destination/dev_null/DevNullBasicFunctionalityIntegrationTest.kt (hunk 8 above)
Comment (johnny-schmidt) [reply airbyte-integrations/connectors/destination-dev-null/src/test-integration/kotlin/io/airbyte/integrations/destination/dev_null/DevNullBasicFunctionalityIntegrationTest.kt:30]: It's not. I assumed I'd gotten out of sync with you at some point and was missing a dependency. I can remove it

File: airbyte-integrations/connectors/destination-dev-null/README.md (hunk 6 above)
Comment (johnny-schmidt) [reply airbyte-integrations/connectors/destination-dev-null/README.md:1]: Yes, I haven't picked up your changes yet, I'm going to just take your readme.

File: airbyte-integrations/connectors/destination-dev-null/build.gradle (hunk 7 above)
Comment (johnny-schmidt) [reply airbyte-integrations/connectors/destination-dev-null/build.gradle:25]: I copied that over from the example and misread it as a core dependency; source implementations use it. we don't need it.

File: airbyte-integrations/connectors/destination-dev-null/src/main/kotlin/io/airbyte/integrations/destination/dev_null/DevNullSpecification.kt [hunk 9]
```diff
@@ -170,20 +180,64 @@ data class RandomSamplingConfig(
...
+/** Non-client-facing configuration. */
+@Singleton
+class DevNullSpecificationInternal : DestinationSpecificationInternal {
+    override val supportedSyncModes =
+        listOf(
+            DestinationSyncMode.OVERWRITE,
+            DestinationSyncMode.APPEND,
+            DestinationSyncMode.APPEND_DEDUP,
+        )
+    override val supportsIncremental = true
```
[the remark this answers is not in the document]
Comment (johnny-schmidt) [reply airbyte-integrations/connectors/destination-dev-null/src/main/kotlin/io/airbyte/integrations/destination/dev_null/DevNullSpecification.kt:242]: if it's meaningless but needs to be set we can always set it to a default in the cdk and ignore it

File: airbyte-integrations/connectors/destination-dev-null/README.md (hunk 6 above)
Comment (johnny-schmidt) [reply airbyte-integrations/connectors/destination-dev-null/README.md:1]: https://github.com/airbytehq/airbyte/pull/45704#discussion_r1769135880

**Comment by johnny-schmidt** [comment]
Merging this down into https://github.com/airbytehq/airbyte/pull/45350 so I can start squashing and rebasing onto @stephane-airbyte 's change. If you. have more comments here I can incorporate them into the final PR

**Pull request merged**